presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java

```diff
@@ -152,15 +153,24 @@ public static Object evaluateConstantExpression(Expression expression, Type expe
         IdentityHashMap<Expression, Type> coercions = new IdentityHashMap<>();
         coercions.putAll(analyzer.getExpressionCoercions());
         coercions.put(expression, expectedType);
-        return evaluateConstantExpression(expression, coercions, metadata, session);
+        return evaluateConstantExpression(expression, coercions, metadata, session, columnReferences);
     }
 
-    public static Object evaluateConstantExpression(Expression expression, IdentityHashMap<Expression, Type> coercions, Metadata metadata, Session session)
+    public static Object evaluateConstantExpression(Expression expression, IdentityHashMap<Expression, Type> coercions, Metadata metadata, Session session, Set<Expression> columnReferences)
     {
         // verify expression is constant
         expression.accept(new DefaultTraversalVisitor<Void, Void>()
         {
             @Override
+            protected Void visitDereferenceExpression(DereferenceExpression node, Void context)
+            {
+                if (columnReferences != null && columnReferences.contains(node)) {
+                    throw new SemanticException(EXPRESSION_NOT_CONSTANT, expression, "Constant expression cannot contain column references");
+                }
```

Comment:
You need to visit the base in the case that it's a row field reference. Otherwise `a.col0` wouldn't throw an exception if `a` is a row type column